Overmind refactor round 3: Run with Overmind

christianalfoni opened the pull request:
Check previous PR for comments: https://github.com/codesandbox/codesandbox-client/pull/2280

Now components can consume **Overmind** instead of **Cerebral** with the url query of `?overmind=true`. The next step now is to ensure that all the pages and logic runs.

## How to contribute

1. Run the project, cerebral devtools and overmind devtools
2. Fire up two different browsers
3. On the browser you want to run Overmind add "overmind" "true" to localStorage
4. Run some logic (click around) on **cerebral** so that you have information in the devtools about how it is run
5. Now do the same on the **overmind** tab. If something breaks, check the Overmind devtools what differs

## Progress

### Root

- [ ] ~~appUnmounted~~
- [x] searchMounted
- [x] sandboxPageMounted
- [x] cliMounted
- [x] cliInstructionsMounted
- [x] githubPageMounted
- [x] connectionChanged
- [x] modalOpened
- [x] modalClosed
- [x] signInClicked
- [x] signInCliClicked
- [x] userMenuOpened
- [x] userMenuClosed
- [x] notificationAdded
- [x] notificationRemoved
- [ ] signInZeitClicked
- [ ] signOutZeitClicked
- [ ] ~~authTokenRequested~~
- [ ] ~~requestAuthorisation~~
- [x] signInGithubClicked
- [x] signOutClicked
- [x] signOutGithubIntegration
- [x] setUpdateStatus
- [x] refetchSandboxInfo
- [x] track

### dashboard

- [x] dashboardMounted
- [x] sandboxesSelected
- [x] setTrashSandboxes
- [x] dragChanged
- [x] orderByChanged
- [x] blacklistedTemplateAdded
- [x] blacklistedTemplateRemoved
- [x] blacklistedTemplatesCleared
- [x] blacklistedTemplatesChanged
- [x] searchChanged
- [x] createSandboxClicked

### deployment

- [x] deployWithNetlify
- [x] getNetlifyDeploys
- [ ] getDeploys
- [ ] deployClicked
- [ ] deploySandboxClicked
- [ ] setDeploymentToDelete
- [ ] deleteDeployment
- [ ] aliasDeployment

### editor

- [x] addNpmDependency
- [x] npmDependencyRemoved
- [x] sandboxChanged
- [x] contentMounted
- [x] resizingStarted
- [x] resizingStopped
- [x] codeSaved
- [x] codeChanged
- [x] saveClicked
- [x] createZipClicked
- [x] forkSandboxClicked
- [x] likeSandboxToggled
- [x] moduleSelected
- [x] clearModuleSelected
- [x] moduleDoubleClicked
- [x] tabClosed
- [x] tabMoved
- [x] prettifyClicked
- [x] errorsCleared
- [x] projectViewToggled
- [x] previewActionReceived
- [x] ~~privacyUpdated~~
- [x] frozenUpdated
- [x] ~~quickActionsOpened~~
- [x] ~~quickActionsClosed~~
- [x] ~~setPreviewContent~~
- [x] togglePreviewContent
- [x] currentTabChanged
- [x] discardModuleChanges
- [x] fetchEnvironmentVariables
- [x] updateEnvironmentVariables
- [x] deleteEnvironmentVariable
- [x] toggleEditorPreviewLayout
- [x] onNavigateAway

### explore

- [x] popularSandboxesMounted
- [x] pickSandbox
- [x] pickSandboxModal
- [ ] ~~getSandbox~~
- [ ] ~~pickedSandboxesMounted~~

### files

- [x] addedFileToSandbox
- [x] gotUploadedFiles
- [x] deletedUploadedFile
- [x] filesUploaded
- [x] moduleCreated
- [x] moduleRenamed
- [x] directoryCreated
- [x] directoryRenamed
- [x] moduleMovedToDirectory
- [x] directoryMovedToDirectory
- [x] directoryDeleted
- [x] moduleDeleted
- [x] createModulesByPath
- [x] syncSandbox
- [x] removeModule
- [x] removeDirectory

### git

- [x] repoTitleChanged
- [x] createRepoClicked
- [x] gitMounted
- [x] createCommitClicked
- [x] subjectChanged
- [x] descriptionChanged
- [x] createPrClicked

### live

- [x] roomJoined
- [x] createLiveClicked
- [x] liveMessageReceived
- [x] onTransformMade
- [x] applyTransformation
- [ ] onCodeReceived
- [x] onOperationApplied
- [x] onSelectionChanged
- [x] onSelectionDecorationsApplied
- [x] onModeChanged
- [x] onAddEditorClicked
- [x] onRemoveEditorClicked
- [x] onSessionCloseClicked
- [x] onNavigateAway
- [x] onToggleNotificationsHidden
- [x] onSendChat
- [x] onChatEnabledChange
- [x] onFollow
- [ ] onModuleStateMismatch

### patron

- [x] patronMounted
- [x] priceChanged
- [x] createSubscriptionClicked
- [x] updateSubscriptionClicked
- [x] cancelSubscriptionClicked
- [x] tryAgainClicked

### preferences

- [ ] ~~viewModeChanged~~
- [ ] ~~devtoolsToggled~~
- [ ] setDevtoolsOpen - *Can not figure out how this works?*
- [x] itemIdChanged
- [x] settingChanged
- [x] setBadgeVisibility
- [x] paymentDetailsRequested
- [x] paymentDetailsUpdated
- [x] keybindingChanged
- [x] zenModeToggled
- [x] codeMirrorForced

### profile

- [x] profileMounted
- [x] sandboxesPageChanged
- [x] likedSandboxesPageChanged
- [x] selectSandboxClicked
- [x] newSandboxShowcaseSelected
- [x] deleteSandboxClicked
- [x] sandboxDeleted

### server

- [x] restartSandbox
- [x] restartContainer
- [ ] ~~statusChanged~~
- [ ] ~~containerStatusChanged~~
- [x] onSSEMessage
- [x] onCodeSandboxAPIMessage
- [x] onBrowserFromPortOpened

### user-notifications

- [x] notificationsOpened
- [x] notificationsClosed
- [ ] **messageReceived** - *Oskar has an extra user, try by inviting to team*

### workspace

- [x] valueChanged
- [x] tagChanged
- [x] tagAdded
- [x] tagRemoved
- [x] sandboxInfoUpdated
- [x] externalResourceAdded
- [x] externalResourceRemoved
- [ ] ~~integrationsOpened~~
- [x] sandboxDeleted
- [x] sandboxPrivacyChanged
- [x] setWorkspaceItem
- [x] toggleCurrentWorkspaceItem
- [x] setWorkspaceHidden

MichaelDeBoey commented on packages/app/src/app/overmind/factories.ts:
```diff
@@ -61,3 +61,26 @@ export function withLoadApp<T>(
+    ) {
+      await actions.editor.internal.forkSandbox(state.editor.currentId);
+
+      return continueAction(context, payload);
```
This one isn't necessary since you're doing exactly the same in the else

christianalfoni replied:
Ops :) Thanks!

MichaelDeBoey commented on packages/app/src/app/overmind/factories.ts:
```diff
@@ -61,3 +61,26 @@ export function withLoadApp<T>(
+export function withOwnedSandbox<T>(
+  continueAction: AsyncAction<T>
+): AsyncAction<T> {
+  return async (context, payload) => {
```
If you make this an arrow function, this can be an implicit return.

```ts
export const withOwnedSandbox = async <T>(
  continueAction: AsyncAction<T>
): AsyncAction<T> => async (context, payload) => {
  // ...
};
```

christianalfoni replied:
Jup, personally I think it reduces readability a bit, but we have a linting rule on this... fixed :)

MichaelDeBoey commented on packages/app/src/app/overmind/namespaces/editor/actions.ts:
```diff
@@ -167,41 +166,41 @@ export const codeChanged: Action<{
+      // might have changed some modules waiting for saving
+      currentlyChangedModuleShortids.forEach(moduleShortid => {
+        if (!state.editor.changedModuleShortids.includes(moduleShortid)) {
+          state.editor.changedModuleShortids.push(moduleShortid);
```
Don't mutate the `state.editor.changedModuleShortids` array, give it a new value instead.

```ts
state.editor.changedModuleShortids =  [...state.editor.changedModuleShortids, moduleShortid];
```

christianalfoni replied:
With Overmind we mutate as it tells the components exactly what we did, improving performance :)

christianalfoni replied:
Not that it practically matters here really, as a **push**, **splice** etc. also causes a "change to the array". But just wanted to mention that we do not depend on reference checking (immutability) in either Cerebral (mobx) or Overmind :)

MichaelDeBoey replied:
It's more a best practice to not mutate variables, but instead change the total value instead.
This _can_ cause bugs when we do mutate.

christianalfoni replied:
Mutations in general is bad, but the state of Overmind is like an API. Everything is controlled and tracked :) This is why Redux has performance issues in complex UIs. It is unable to pinpoint exact changes, as any change causes change to all parent objects, meaning all components looking at state at any level of the same hierarchy will need to reconcile.

I really like the simplicity of immutability and reference checking for React, but it just does not work in complex UIs as it causes performance issues without a lot of manual props checking, normalizing state and stuff.

Mobx / Mobx-State-Tree also uses controlled mutability. I think fundamentally Redux (immutability) is doing the right thing, but the API is hell to work with as JavaScript has no immutable native API and JavaScript/React is not fast enough to manage a full reconcile on any change. You have to do a lot of React.memo etc. to manage performance.

Wow, that got me going! :) Anyways, this specific change does not matter for performance, except that you will not see in the devtools what you actually pushed to the array as that information is lost.

MichaelDeBoey replied:
I understand the reasoning behind it 🙂

MichaelDeBoey commented on packages/app/src/app/pages/Profile/Sandboxes/index.js:
```diff
@@ -68,10 +83,14 @@ class Sandboxes extends React.Component {
+      !sandboxes ||
+      !this.getSandboxesByPage(sandboxes, page)
+    )
       return <div />;
```
Put curly braces around condition body

christianalfoni replied:
fixed :)

MichaelDeBoey commented on packages/app/src/app/pages/Sandbox/Editor/Workspace/Files/DirectoryEntry/DirectoryChildren/index.js:
```diff
@@ -56,6 +56,7 @@ class DirectoryChildren extends React.Component {
               key={dir.id}
               siblings={[...directories, ...modules]}
               depth={depth + 1}
+              signals={this.props.signals}
```
Don't pass `signals` to `DirectoryEntry`, wrap it with an `inject` instead

christianalfoni replied:
We have to pass it due to the temporary wrapper. With Overmind the inject actually does nothing. It is the observer that provides this stuff, though that can not be around the Drag HOC. Drag is a messy thing in React, this should be refactored with hooks later.

MichaelDeBoey replied:
Hmmm too bad it needs to be like this.
Maybe add an `TODO` next to it, so we don't forget about it? 🤔

christianalfoni replied:
Jup, added! :)

MichaelDeBoey commented on packages/app/src/app/pages/Sandbox/Editor/Workspace/Files/index.js:
```diff
@@ -43,6 +43,7 @@ class Files extends React.Component {
         root
         getModulePath={this.getModulePath}
         title={sandbox.title || 'Project'}
+        signals={this.props.signals}
```
Don't pass `signals` to `DirectoryEntry`, wrap it with an `inject` instead

christianalfoni replied:
We have to pass it due to the temporary wrapper. With Overmind the **inject** actually does nothing. It is the observer that provides this stuff, though that can not be around the Drag HOC. Drag is a messy thing in React, this should be refactored with hooks later.

MichaelDeBoey replied:
Hmmm too bad it needs to be like this.
Maybe add an `TODO` next to it, so we don't forget about it? 🤔

christianalfoni replied:
Jup :)

CompuIves commented:
>  blacklistedTemplateAdded - Can not find this component?
  blacklistedTemplateRemoved - Can not find this component?
  blacklistedTemplatesCleared - Can not find this component?
  blacklistedTemplatesChanged - Can not find this component?

Found in the dashboard when you filter on templates.

> saveClicked

Might be here, but only called from iPad?
https://github.com/CompuIves/codesandbox-client/blob/11f13c7960f6f62a2832d0ff2710e602433ea88b/packages/app/src/app/store/modules/editor/utils.js#L3-L23

> toggleEditorPreviewLayout

Think this one is still used in VSCode (you can do CMD+Shift+P -> "Flip Preview Layout"): https://github.com/CompuIves/codesandbox-client/blob/11f13c7960f6f62a2832d0ff2710e602433ea88b/packages/app/src/app/vscode/index.ts#L54-L69

MichaelDeBoey commented on packages/app/src/app/overmind/internalActions.ts:
```diff
@@ -258,7 +258,10 @@ export const setCurrentSandbox: AsyncAction<Sandbox> = async (
   effects.fsSync.syncCurrentSandbox();
-  effects.router.updateSandboxUrl(sandbox);
+
+  if (oldSandboxId && oldSandboxId !== state.editor.currentId) {
```
No need to check if `oldSandboxId` exists here.

christianalfoni replied:
Actually we do :) We only want to check if they are different if there actually existed an old id. Basically, we do not want to check if they are different if it is the first sandbox you are opening. It could most certainly be expressed better :)

christianalfoni replied:
This thing has actually been removed :)

MichaelDeBoey commented on packages/app/src/app/index.js:
```diff
@@ -166,6 +166,11 @@ async function initialize() {
+            window.opener && window.opener !== window
+              ? false
+              : 'localhost:3031',
+          name: 'Codesandbox',
```
```suggestion
          name: 'CodeSandbox',
```

MichaelDeBoey commented on packages/app/src/app/index.js:
```diff
@@ -166,6 +166,11 @@ async function initialize() {
 
         overmind = createOvermind(config, {
+          devtools:
+            window.opener && window.opener !== window
```
No need to check if `window.opener` exists here

christianalfoni replied:
We only want to check if there actually is a `window.opener` there. Or it would give "true" when there is no opener, which would cause the wrong result :)

MichaelDeBoey commented on packages/app/src/app/overmind/effects/live/index.ts:
```diff
@@ -80,21 +85,32 @@ export default {
-    channel = socket.getSocket().channel(`live:${roomId}`, {});
+  joinChannel(
+    roomId: string
+  ): Promise<{
```
I think it's best to extract this in a separate type so you don't need the `as any`, but can do `as type`.

christianalfoni replied:
Yeah, I fixed it, but for Typescript it does not actually matter. If you set an explicit return type on a method, that return type will be used even if you resolve/return an `any` :)

MichaelDeBoey commented on packages/app/src/app/overmind/namespaces/live/actions.ts:
```diff
@@ -401,8 +402,11 @@ export const onTransformMade: Action<{
+    console.log(client);
+    client.applyClient(operation);
   } catch (e) {
+    console.log(e);
```
Don't forget to delete this one.

christianalfoni replied:
Ops, yeah, thanks :)

MichaelDeBoey commented on packages/app/src/app/overmind/namespaces/live/actions.ts:
```diff
@@ -401,8 +402,11 @@ export const onTransformMade: Action<{
   try {
-    effects.live.getClient(moduleShortid).applyClient(operation);
+    const client = effects.live.getClient(moduleShortid);
+    console.log(client);
```
Don't forget to delete this one.

christianalfoni replied:
Jup jup, thanks :)

MichaelDeBoey commented on packages/app/src/app/index.js:
```diff
@@ -159,7 +159,15 @@ async function initialize() {
+  };
+
   if (localStorage.getItem('overmind') === 'true') {
+    console.log('########');
```
Don't forget to delete this one.

christianalfoni replied:
Ops, removed :)

christianalfoni commented:
@CompuIves Went through:

>  blacklistedTemplateAdded - Can not find this component?
  blacklistedTemplateRemoved - Can not find this component?
  blacklistedTemplatesCleared - Can not find this component?
  blacklistedTemplatesChanged - Can not find this component?

Found in the dashboard when you filter on templates.

**Tested and fixed**

> createSandboxClicked - Can not figure out when this is possible?

When you create a sandbox from a folder in your dashboard, we fork it and set the folder to the current folder. Used here: https://github.com/CompuIves/codesandbox-client/blob/11f13c7960f6f62a2832d0ff2710e602433ea88b/packages/app/src/app/pages/Dashboard/Content/CreateNewSandbox/index.js#L44

**Tested and fixed**

> saveClicked

Might be here, but only called from iPad?
https://github.com/CompuIves/codesandbox-client/blob/11f13c7960f6f62a2832d0ff2710e602433ea88b/packages/app/src/app/store/modules/editor/utils.js#L3-L23

**Waiting with iPad in case iOS 13 gets out before we need to fix it**

> toggleEditorPreviewLayout

Think this one is still used in VSCode (you can do CMD+Shift+P -> "Flip Preview Layout"): https://github.com/CompuIves/codesandbox-client/blob/11f13c7960f6f62a2832d0ff2710e602433ea88b/packages/app/src/app/vscode/index.ts#L54-L69

**Tested and fixed**

> files.addedFileToSandbox

Seems in the storage modal? https://github.com/CompuIves/codesandbox-client/blob/11f13c7960f6f62a2832d0ff2710e602433ea88b/packages/app/src/app/pages/common/Modals/StorageManagementModal/index.js#L45-L48

**Though there is no UI for adding any files in this modal? Only to review existing uploads?**

CompuIves commented on packages/app/src/app/overmind/namespaces/files/internalActions.ts:
```diff
@@ -93,8 +93,6 @@ export const uploadFiles: AsyncAction<
             /\.pug$/.test(filePath) ||
             /\.svg$/.test(filePath) ||
-            /\.md$/.test(filePath) ||
-            /\.svelte$/.test(filePath) ||
```
I think this one should be added again :D

christianalfoni replied:
Added :)
